File: client/webauthn-client-flow/src/commonMain/kotlin/dev/webauthn/client/PasskeyFlow.kt
@file:Suppress("UndocumentedPublicFunction", "UndocumentedPublicProperty")

package dev.webauthn.client

import dev.webauthn.model.PublicKeyCredentialCreationOptions
import dev.webauthn.model.PublicKeyCredentialRequestOptions
import dev.webauthn.model.RawAuthenticationResponse
import dev.webauthn.model.RawRegistrationResponse
import kotlinx.coroutines.sync.Mutex

/** Server-owned continuation state and options returned when a ceremony starts. */
public data class CeremonyStart<State, Options>(
    public val state: State,
    public val options: Options,
)

/** Backend contract for a registration ceremony with opaque state and application output. */
public interface RegistrationBackend<Input, State, Output> {
    public suspend fun start(input: Input): CeremonyStart<State, PublicKeyCredentialCreationOptions>

    public suspend fun finish(state: State, response: RawRegistrationResponse): Output
}

/** Backend contract for an authentication ceremony with opaque state and application output. */
public interface AuthenticationBackend<Input, State, Output> {
    public suspend fun start(input: Input): CeremonyStart<State, PublicKeyCredentialRequestOptions>

    public suspend fun finish(state: State, response: RawAuthenticationResponse): Output
}

/** Completion value or deliberately classified platform/concurrency failure. */
public sealed interface CeremonyResult<out Output> {
    public data class Success<Output>(public val value: Output) : CeremonyResult<Output>

    public data class Failure(public val error: CeremonyFailure) : CeremonyResult<Nothing>
}

/** Failures that the generic flow can classify without owning application/backend semantics. */
public sealed interface CeremonyFailure {
    /** Another ceremony is already using this flow instance. */
    public data object AlreadyInProgress : CeremonyFailure

    /** The platform client returned a deliberate passkey failure. */
    public data class Platform(public val error: PasskeyClientError) : CeremonyFailure
}

/**
 * Stateless ceremony orchestration that carries opaque backend state from start through finish.
 * Backend and callback exceptions deliberately propagate to the application.
 */
public class PasskeyFlow(
    private val passkeyClient: PasskeyClient,
) {
    private val ceremonyMutex = Mutex()

    public suspend fun <Input, State, Output> register(
        input: Input,
        backend: RegistrationBackend<Input, State, Output>,
        onPhaseChanged: (PasskeyPhase) -> Unit = {},
    ): CeremonyResult<Output> {
        return runCeremony(
            start = { backend.start(input) },
            prompt = passkeyClient::createCredential,
            finish = backend::finish,
            onPhaseChanged = onPhaseChanged,
        )
    }

    public suspend fun <Input, State, Output> signIn(
        input: Input,
        backend: AuthenticationBackend<Input, State, Output>,
        onPhaseChanged: (PasskeyPhase) -> Unit = {},
    ): CeremonyResult<Output> {
        return runCeremony(
            start = { backend.start(input) },
            prompt = passkeyClient::getAssertion,
            finish = backend::finish,
            onPhaseChanged = onPhaseChanged,
        )
    }

    private suspend fun <State, Options, Response, Output> runCeremony(
        start: suspend () -> CeremonyStart<State, Options>,
        prompt: suspend (Options) -> PasskeyResult<Response>,
        finish: suspend (State, Response) -> Output,
        onPhaseChanged: (PasskeyPhase) -> Unit,
    ): CeremonyResult<Output> {
        if (!ceremonyMutex.tryLock()) {
            return CeremonyResult.Failure(CeremonyFailure.AlreadyInProgress)
        }
        try {
            onPhaseChanged(PasskeyPhase.STARTING)
            val ceremonyStart = start()
            onPhaseChanged(PasskeyPhase.PLATFORM_PROMPT)
            val response = when (val result = prompt(ceremonyStart.options)) {
                is PasskeyResult.Success -> result.value
                is PasskeyResult.Failure -> return CeremonyResult.Failure(CeremonyFailure.Platform(result.error))
            }
            onPhaseChanged(PasskeyPhase.FINISHING)
            return CeremonyResult.Success(finish(ceremonyStart.state, response))
        } finally {
            ceremonyMutex.unlock()
        }
    }
}

File: swift/Sources/WebAuthnFlow/PasskeyFlow.swift
import Foundation
import WebAuthn

/// Server-owned continuation state and options returned when a ceremony starts.
public struct CeremonyStart<State> {
    /// Opaque backend state forwarded unchanged to the finish operation.
    public let state: State

    /// UTF-8 WebAuthn options JSON passed to the platform client.
    public let optionsJSON: Data

    public init(state: State, optionsJSON: Data) {
        self.state = state
        self.optionsJSON = optionsJSON
    }
}

extension CeremonyStart: Equatable where State: Equatable {}
extension CeremonyStart: Sendable where State: Sendable {}

/// Application-owned backend contract for a registration ceremony.
@MainActor
public protocol RegistrationBackend {
    associatedtype Input
    associatedtype State
    associatedtype Output

    /// Starts registration and returns opaque continuation state plus creation options.
    func start(input: Input) async throws -> CeremonyStart<State>

    /// Finishes registration with the unchanged state and platform response.
    func finish(state: State, responseJSON: Data) async throws -> Output
}

/// Application-owned backend contract for an authentication ceremony.
@MainActor
public protocol AuthenticationBackend {
    associatedtype Input
    associatedtype State
    associatedtype Output

    /// Starts authentication and returns opaque continuation state plus request options.
    func start(input: Input) async throws -> CeremonyStart<State>

    /// Finishes authentication with the unchanged state and platform response.
    func finish(state: State, responseJSON: Data) async throws -> Output
}

/// Lifecycle phase of a passkey ceremony.
public enum PasskeyPhase: Equatable, Sendable {
    case starting
    case platformPrompt
    case finishing
}

/// Failures classified by the application-neutral flow.
public enum CeremonyFailure: Equatable, Sendable {
    /// Another ceremony is already using this flow instance.
    case alreadyInProgress

    /// The platform client returned a typed passkey failure.
    case platform(PasskeyClientError)
}

/// Completion value or a failure classified by the application-neutral flow.
public enum CeremonyResult<Output> {
    case success(Output)
    case failure(CeremonyFailure)
}

extension CeremonyResult: Equatable where Output: Equatable {}
extension CeremonyResult: Sendable where Output: Sendable {}

/// Registration and authentication start-prompt-finish orchestration.
///
/// Backend and phase-callback errors deliberately propagate to the caller. The flow classifies only its own
/// concurrent-use rejection and typed ``PasskeyClientError`` values returned by the platform client.
@MainActor
public final class PasskeyFlow {
    private let client: any PasskeyClientProtocol
    private var ceremonyInProgress = false

    public init(client: any PasskeyClientProtocol) {
        self.client = client
    }

    /// Runs registration while preserving backend state and application-defined output.
    public func register<Backend: RegistrationBackend>(
        _ input: Backend.Input,
        backend: Backend,
        onPhaseChanged: (PasskeyPhase) throws -> Void = { _ in }
    ) async throws -> CeremonyResult<Backend.Output> {
        try await runCeremony(
            start: { try await backend.start(input: input) },
            prompt: { try await self.client.createCredential(optionsJSON: $0) },
            finish: { try await backend.finish(state: $0, responseJSON: $1) },
            onPhaseChanged: onPhaseChanged
        )
    }

    /// Runs authentication while preserving backend state and application-defined output.
    public func signIn<Backend: AuthenticationBackend>(
        _ input: Backend.Input,
        backend: Backend,
        onPhaseChanged: (PasskeyPhase) throws -> Void = { _ in }
    ) async throws -> CeremonyResult<Backend.Output> {
        try await runCeremony(
            start: { try await backend.start(input: input) },
            prompt: { try await self.client.getAssertion(optionsJSON: $0) },
            finish: { try await backend.finish(state: $0, responseJSON: $1) },
            onPhaseChanged: onPhaseChanged
        )
    }

    private func runCeremony<State, Output>(
        start: () async throws -> CeremonyStart<State>,
        prompt: (Data) async throws -> Data,
        finish: (State, Data) async throws -> Output,
        onPhaseChanged: (PasskeyPhase) throws -> Void
    ) async throws -> CeremonyResult<Output> {
        guard !ceremonyInProgress else {
            return .failure(.alreadyInProgress)
        }
        ceremonyInProgress = true
        defer { ceremonyInProgress = false }

        try Task.checkCancellation()
        try onPhaseChanged(.starting)
        let ceremony = try await start()

        try Task.checkCancellation()
        try onPhaseChanged(.platformPrompt)
        let response: Data
        do {
            response = try await prompt(ceremony.optionsJSON)
        } catch let error as PasskeyClientError {
            return .failure(.platform(error))
        }

        try Task.checkCancellation()
        try onPhaseChanged(.finishing)
        return .success(try await finish(ceremony.state, response))
    }
}
